alternator/expressions.cc

```diff
@@ -667,6 +666,55 @@ std::unordered_map<std::string_view, function_handler_type*> function_handlers {
     },
 };
 
+// Given a parsed::path and an item read from the table, extract the value
+// of a certain attribute path, such as "a" or "a.b.c[3]". Returns a null
+// value if the item or the requested attribute does not exist.
+// Note that the item is assumed to be encoded in JSON using DynamoDB
+// conventions - each level of a nested document is a map with one key -
+// a type (e.g., "M" for map) - and its value is the representation of
+// that value.
+static rjson::value extract_path(const rjson::value* item,
+        const parsed::path& p, calculate_value_caller caller) {
+    if (!item) {
+        return rjson::null_value();
+    }
+    const rjson::value* v = rjson::find(*item, p.root());
+    if (!v) {
+        return rjson::null_value();
+    }
+    for (const auto& op : p.operators()) {
+        if (!v->IsObject() || v->MemberCount() != 1) {
+            // This shouldn't happen. We shouldn't have stored malformed
+            // objects. But today Alternator does not validate the structure
+            // of nested documents before storing them, so this can happen on
+            // read.
```

Comment:
A workaround for the client would simply be to delete the item by its key once this error occurs, right?

Reply:
Yes. What we really need to do at some point (unrelated to this issue) is to add some sort of validation that the JSON we insert into an attribute is valid. We do validate that it's legal JSON (we parse that JSON...) but not that it's legal DynamoDB-serialized format. If we do this during write, we won't need to do it during read.

Reply:
Created https://github.com/scylladb/scylla/issues/8070 - but I'm not planning to do that right now.